File: 55-jump-game/jump-game.cpp
class Solution {
public:
    bool canJump(vector<int>& nums) {
        if(nums.size() == 1) return true;
        unordered_map<int, bool> val;
        int lastIdx = nums.size()-1;
        for(int i = nums.size() -2;i>=0;i--){
            for(int j = 1;j<=nums[i];j++){
                if(i+j>=lastIdx || val[i+j]){
                    val[i] = true;
                    break;
                }else{
                    val[i] = false;
                }
            }
        }
        return val[0];
    }
};

File: 55-jump-game/jump-game.py
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        if len(nums) ==1:
            return True
        sum_el = 0

        for i in range(len(nums)):
            if i == 0:
                sum_el = nums[i]
                continue
            if sum_el<i:
                return False
            sum_el = i+ max(sum_el - i, nums[i])
        return True
        